### PR TITLE
fix/clean up waiter model

### PR DESCRIPTION
**Pull request by billyxs**
Documentation was a little off so I made some fixes

There are some half baked ideas of where I thought I wanted to take this. 

I've never implemented prepareWaiter() in a project, and it looked off.  Part of this direction included these extra properties on the model that were never used. Going to remove them for now.

isCanceled was never fully implemented, and I'm not sure it's needed. Was considering removing it as well. clearWaiter and destroyWaiter may be enough, but I'm still having a think about it.

### CODE REVIEW

**Comment by seesleestak** [comment]
One of devs at work had asked me if the waiter did cancelling requests. I'd seen in there and thought about trying to get it fleshed out.

**Comment by billyxs** [comment]
It comes down to "what does it mean to cancel"? Some libraries like axios allow canceling of requests, but that setup is all owned by axios. You would need to pass a cancel handler to waiter to do it for you, but is that a responsibility of the waiter?

In a different space, what does it mean to cancel a Promise? Not listen to it anymore? Delete the request from the store? Clear and destroy do that. How is cancel different? What is it looking to achieve?

File: src/actions.js [hunk 1]
```diff
@@ -73,14 +78,48 @@ export function callWaiter(name, { requestCreator, params }) {
...
+        // A new request was made, should not listen to this request anymore
+        if (preEndState.id !== waiterData.id) {
+          return null;
+        }
+
+        // Request was canceled
+        if (preEndState.isCanceled) {
+          return null;
+        }
+
```
Comment (billyxs) [inline src/actions.js]: This is what I was working on, and wasn't sure if isCanceled was needed.

**Comment by seesleestak** [comment]
> It comes down to "what does it mean to cancel"? Some libraries like axios allow canceling/aborting of requests, but that setup is all owned by axios. You would need to pass a cancel handler to waiter to do it for you, but is that a responsibility of the waiter?
> 
> In a different space, what does it mean to cancel a Promise? Not listen to it anymore? Delete the request from the store? Clear and destroy do that. How is cancel different? What is it looking to achieve?

**Comment by seesleestak** [comment]
Yeah, perhaps its not the domain of this library to handle that. Good stuff to be thinking about, I think.

File: src/actions.js (hunk 1 above)
Comment (billyxs) [reply src/actions.js]: This currently doesn't past tests, and is just in a WIP state. If the waiter is destroyed, it probably doesn't need a check. If the waiter id changes, that could have a strange effect on the model if an earlier request overwrites a later request. Still musing about on it

File: src/actions.js [hunk 2]
```diff
@@ -73,14 +78,48 @@ export function callWaiter(name, { requestCreator, params }) {
...
-    const request = waiterData.requestCreator(params, dispatch);
-    dispatch(initRequest(name, { request, params }));
+    const request = waiterData.requestCreator(waiterData.params, dispatch);
+    dispatch(initRequest(name, { request }));
     request
       .then((data) => {
+        const preEndState = getWaiter(getState(), name);
+
+        // waiter was destroyed
+        if (preEndState.name === null) {
```
Comment (billyxs) [inline src/actions.js]: This probably makes more sense
```
        // waiter was cleared
        if (preEndState.request === null) {
          return preEndState;
        }
```

File: src/actions.js [hunk 3]
```diff
@@ -55,10 +55,43 @@ export const destroyWaiter = (name) => ({
...
+    }
+
+    // A new request was made
+    // should not listen to this request anymore
+    if (waiterId !== waiterData.id) {
+      return waiterData;
+    }
+
+    // Request was canceled
+    if (waiterData.isCanceled) {
```
Comment (billyxs) [inline src/actions.js]: Going to leave this here for now. If you wanted to "flesh" out something more advanced, happy to see it.

Comment (seesleestak) [reply src/actions.js]: Sounds goof

Comment (seesleestak) [reply src/actions.js]: Sounds goof

**Comment by billyxs** [comment]
I think I fixed most of the embarrassing mistakes and inconsistencies in the way the waiter parent and child models are handled. Probably wouldn't notice if you only used the basic lifecycle stuff.

File: src/actions.js [hunk 4]
```diff
@@ -55,10 +55,43 @@ export const destroyWaiter = (name) => ({
...
 
+export function waiterResponseHandler(
+  name,
+  { waiterId, responseAction, responseData }
+) {
+  return (dispatch, getState) => {
+    const waiterData = getWaiter(getState(), name);
+
+    // waiter was cleared or destroyed
+    if (waiterData.request === null) {
```
Comment (seesleestak) [inline src/actions.js]: Would this condition only happen when the waiter is cleared and _not_ destroyed? If the waiter had been destroyed, wouldn’t we get `undefined` here?

File: src/waiter-reducer.js [hunk 5]
```diff
@@ -112,6 +110,7 @@ const reducerMap = {
     elapsedTime: getTime() - state.startTime,
     lastModified: getTime(),
   }),
+
   [t.CLEAR]: (state) => ({
     ...state,
```
Comment (billyxs) [inline src/waiter-reducer.js]: Was wondering about params here. clearWaiter is about clearing the request lifecycle. The params and requestCreator should be available to use again.

File: README.md [hunk 6]
```diff
@@ -207,10 +205,21 @@ used again.
...
+### clearAll()
+
+Clear all the waiters in the store.
+
+```javascript
+import { clearAll } from 'redux-waiter';
+
+// In you redux environment
+dispatch(clearAll());
+```
```
[the remark this answers is not in the document]
Comment (billyxs) [reply README.md]: sorry, missed this comment. Yes we do. Thanks!

File: src/index.js [hunk 7]
```diff
@@ -21,6 +21,8 @@ export const callWaiter = actions.callWaiter;
 export const clearWaiter = actions.clearWaiter;
 export const prepareWaiter = actions.prepareWaiter;
 export const destroyWaiter = actions.destroyWaiter;
+export const clearAll = actions.clearAll;
+export const destroyAll = actions.destroyAll;
```
Comment (seesleestak) [inline src/index.js]: Not sure if you're interested, but I think we could have some more concise syntax here with destructuring:
```
export const {
  callWaiter,
  clearWaiter,
  prepareWaiter,
  destroyWaiter,
  clearAll,
  destroyAll,
} = actions
```
Just an alternative :)

**Comment by seesleestak** [comment]
Love having all the tests. Feels good.

File: src/index.js (hunk 7 above)
Comment (billyxs) [reply src/index.js]: sounds good 👍

**Pull request merged**